### PR TITLE
Adding object wrappers for SIMD types.

### PR DESCRIPTION
**Pull request by arunetm**
Adding object wrappers for SIMD types.

1. Adding SIMD object wrappers for all SIMD types.
   Populated SIMD object prototype with the needed methods (toString,
   toLocaleString, valueOf etc)
2. Added value semantics for SIMD types.
   SIMD variables can be used in operations similar to the other
   javascript variables.
3. BugFix: Chakra could become unresponsive if @@ToPrimitive of a
   javascript object is overridden by null or undefined. Adding fix.
4. Enabling previosuly disabled type conversion operationfor Uint32x4.
   Adding tests.

[Review on Reviewable](https://reviewable.io/reviews/microsoft/chakracore/325)

### CODE REVIEW

**Comment by arunetm** [comment]
@dotnet-bot test this please

**Comment by arunetm** [comment]
@abchatra @LouisLaf please review this feature.

**Comment by arunetm** [comment]
Thank you for the feedback. I am working on a new commit with the changes and the following fixes.
1. SIMDVariable.valueOf will be modified to return the SIMD value as per the spec.
2. New use cases and tests for valueOf, toString, adding new operations on simd types etc.

**Comment by arunetm** [comment]
Updating PR with new commit. 
1. Incorporating review feedback.
2. Bug fix for valueOf operation on SIMD types. 
3. More unittests and use cases. 
4. Code refactoring.

**Comment by arunetm** [comment]
@dotnet-bot test this please

**Comment by arunetm** [comment]
@dilijev Can you please help to sign off the PR. Thanks.

**Comment by dilijev** [comment]
Reviewed 65 of 65 files at r1.
Review status: all files reviewed at latest revision, 26 unresolved discussions, some commit checks failed.

---

*[lib/Runtime/Library/JavascriptLibrary.cpp, line 328 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KBy7j5jTkpjp-VnEpEQ:-KBy7j5jTkpjp-VnEpER:-1551665753) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/lib/Runtime/Library/JavascriptLibrary.cpp#L328)):*
nit: trailing whitespace

---

*[lib/Runtime/Library/JavascriptLibrary.cpp, line 714 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KBy7m40tSFRjouT-qZJ:-KBy7m40tSFRjouT-qZK:199826290) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/lib/Runtime/Library/JavascriptLibrary.cpp#L714)):*
nit: trailing whitespace (here and slightly above)

---

*[lib/Runtime/Library/JavascriptSimdInt16x8.h, line 36 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByCDpWcjaVvisqUwWz:-KByCDpX5u9xxN1TdiSr:-1551665753) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/lib/Runtime/Library/JavascriptSimdInt16x8.h#L36)):*
nit: trailing whitespace

---

*[lib/Runtime/Library/JavascriptSimdInt32x4.h, line 20 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByCIc2gVNZaToQ-8gF:-KByCIc2gVNZaToQ-8gG:-1551665753) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/lib/Runtime/Library/JavascriptSimdInt32x4.h#L20)):*
nit: trailing whitespace

---

*[lib/Runtime/Library/JavascriptSimdInt8x16.h, line 43 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByCNySI4_0TicCGFVM:-KByCNySI4_0TicCGFVN:-1551665753) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/lib/Runtime/Library/JavascriptSimdInt8x16.h#L43)):*
nit: trailing whitespace

---

*[Lib/Runtime/Library/JavascriptSimdObject.cpp, line 189 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByCU_pZ48nbSXNq8Dl:-KByCU_pZ48nbSXNq8Dm:-651728003) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/Lib/Runtime/Library/JavascriptSimdObject.cpp#L189)):*
nit: trailing whitespace here and above

---

*[Lib/Runtime/Library/JavascriptSimdObject.cpp, line 257 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByCXVabhhc6GgDeXU0:-KByCXVabhhc6GgDeXU1:-797135463) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/Lib/Runtime/Library/JavascriptSimdObject.cpp#L257)):*
nit: trailing whitespace. 

---

*[Lib/Runtime/Library/JavascriptSimdObject.h, line 24 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByCaKlIhS60oV7CNPH:-KByCaKlIhS60oV7CNPI:-1551665753) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/Lib/Runtime/Library/JavascriptSimdObject.h#L24)):*
nit: trailing whitespace

---

*[Lib/Runtime/Library/JavascriptSimdType.cpp, line 16 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByCdoIpDD2KMIprktw:-KByCdoIpDD2KMIprktx:109580105) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/Lib/Runtime/Library/JavascriptSimdType.cpp#L16)):*
nit: extra blank lines

---

*[Lib/Runtime/Library/JavascriptSimdType.cpp, line 172 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByChFVWQUFmZtASzdZ:-KByChFVWQUFmZtASzd_:1491680145) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/Lib/Runtime/Library/JavascriptSimdType.cpp#L172)):*
nit : extra blank lines

---

*[Lib/Runtime/Library/JavascriptSimdType.cpp, line 196 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByCjyjDLgfNnX_5Ylz:-KByCjyjDLgfNnX_5Ym-:-412107158) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/Lib/Runtime/Library/JavascriptSimdType.cpp#L196)):*
nit: extra blank line

---

*[Lib/Runtime/Library/JavascriptSimdType.cpp, line 219 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByClVPkgR9g5C0Kt0B:-KByClVPkgR9g5C0Kt0C:38123079) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/Lib/Runtime/Library/JavascriptSimdType.cpp#L219)):*
nit : unneeded blank line before closing brace

---

*[Lib/Runtime/Library/JavascriptSimdType.h, line 33 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByCpYPt9FzdyE9-CN9:-KByCpYPt9FzdyE9-CNA:-1551665753) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/Lib/Runtime/Library/JavascriptSimdType.h#L33)):*
nit: trailing whitespace

---

*[test/Array/memset_simd.js, line 21 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByD64VWFSjqL3Nt2n4:-KByD64VWFSjqL3Nt2n5:51281430) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/test/Array/memset_simd.js#L21)):*
Wrap this if-statement body in braces.

nits: include space in `if (`, only need one space before a comment token, space after the comment token.

Also applies to the change below.

---

*[test/SIMD.wrappers/rlexe.xml, line 62 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByDT-7nuvRmKKqWKle:-KByDT-7nuvRmKKqWKlf:-1551665753) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/test/SIMD.wrappers/rlexe.xml#L62)):*
nit: trailing whitespace

---

*[test/SIMD.wrappers/testBool16x8.js, line 23 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByDd65j03Q4igmH9kP:-KByDd65j03Q4igmH9kQ:1908654473) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/test/SIMD.wrappers/testBool16x8.js#L23)):*
nit: there is trailing whitespace like this throughout this file. please trim all trailing whitespace

---

*[test/SIMD.wrappers/testBool16x8.js, line 305 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByDXVwaxH1JxeJaATO:-KByDXVwaxH1JxeJaATP:-751275398) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/test/SIMD.wrappers/testBool16x8.js#L305)):*
There are some strange bytes at the end of this line? Make sure there's a blank line at the end of the file.

---

*[test/SIMD.wrappers/testBool32x4.js, line 305 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByDktn0vQbUkUfiYGC:-KByDktn0vQbUkUfiYGD:-324854047) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/test/SIMD.wrappers/testBool32x4.js#L305)):*
strange bytes and trailing whitespace throughout this file too

---

*[test/SIMD.wrappers/testBool8x16.js, line 305 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByDqCgU2hubTB8pr1E:-KByDqChQu076_TYneJ9:-324854047) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/test/SIMD.wrappers/testBool8x16.js#L305)):*
strange bytes and trailing whitespace throughout this file too

---

*[test/SIMD.wrappers/testFloat32x4.js, line 305 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByDu6FakirF6bGmN64:-KByDu6FakirF6bGmN65:678021867) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/test/SIMD.wrappers/testFloat32x4.js#L305)):*
same

---

*[test/SIMD.wrappers/testInt16x8.js, line 309 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByDwlNzyu5bOtYATbt:-KByDwlNzyu5bOtYATbu:678021867) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/test/SIMD.wrappers/testInt16x8.js#L309)):*
same

---

*[test/SIMD.wrappers/testInt32x4.js, line 305 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByDydTwudFEjD7LnVw:-KByDydTwudFEjD7LnVx:678021867) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/test/SIMD.wrappers/testInt32x4.js#L305)):*
same

---

*[test/SIMD.wrappers/testInt8x16.js, line 318 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByE-P6kyYtPtQU9g_b:-KByE-P6kyYtPtQU9g_c:678021867) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/test/SIMD.wrappers/testInt8x16.js#L318)):*
same

---

*[test/SIMD.wrappers/testUint16x8.js, line 309 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByE2Bm4Yx0qy5RNUqu:-KByE2Bm4Yx0qy5RNUqv:678021867) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/test/SIMD.wrappers/testUint16x8.js#L309)):*
same

---

*[test/SIMD.wrappers/testUint32x4.js, line 305 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByE3tYaqEdQd5YEYCi:-KByE3tYaqEdQd5YEYCj:678021867) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/test/SIMD.wrappers/testUint32x4.js#L305)):*
same

---

*[test/SIMD.wrappers/testUint8x16.js, line 317 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByE5Zxdk3hfJ_UhEAL:-KByE5Zxdk3hfJ_UhEAM:678021867) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/test/SIMD.wrappers/testUint8x16.js#L317)):*
same

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/325)*

**Comment by dilijev** [comment]
Reviewed 24 of 26 files at r2.
Review status: all files reviewed at latest revision, 27 unresolved discussions.

---

*[Lib/Runtime/Library/JavascriptSimdObject.cpp, line 189 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KByCU_pZ48nbSXNq8Dl:-KC1YThETyL4SfB7UD-T:-393271004) ([raw file](https://github.com/microsoft/chakracore/blob/ac616c80010c06cab3b44d9a4ace839cb0041bde/Lib/Runtime/Library/JavascriptSimdObject.cpp#L189)):*
Still trailing whitespace in this block.

---

*[test/SIMD.wrappers/testBool8x16.js, line 154 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/325#-KC1YjwJiMQVwN0FRLiR:-KC1YjwJiMQVwN0FRLiS:416498012) ([raw file](https://github.com/microsoft/chakracore/blob/6f97c51000445f5f7c4c467a355b8e8451716522/test/SIMD.wrappers/testBool8x16.js#L154)):*
nit: still space here

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/325)*

**Comment by dilijev** [comment]
The new comments I made about spaces might have been existing before your change. Don't worry about them.

LGTM. Signed off.

File: lib/Runtime/Language/SimdUtils.h [hunk 1]
```diff
@@ -144,6 +144,8 @@ CompileAssert(sizeof(SIMDValue) == 16);
 class ValueType;
 
 namespace Js {
+
+    bool IsSimdType(Var aVar);
```
Comment (abchatra) [inline lib/Runtime/Language/SimdUtils.h]: All the methods here should be in its own class not on global namespace

File: lib/Runtime/Language/JavascriptConversion.cpp [hunk 2]
```diff
@@ -332,6 +339,11 @@ namespace Js
     //----------------------------------------------------------------------------
     Var JavascriptConversion::ToPrimitive(Var aValue, JavascriptHint hint, ScriptContext * requestContext)
     {
+        if (IsSimdType(aValue))
+        {
+            return aValue;
+        }
```
Comment (abchatra) [inline lib/Runtime/Language/JavascriptConversion.cpp]: Can you please move this under default? I am not why this should be the first case as other primitive types are more common

File: lib/Runtime/Language/JavascriptConversion.cpp [hunk 3]
```diff
@@ -264,6 +265,12 @@ namespace Js
     BOOL JavascriptConversion::ToObject(Var aValue, ScriptContext* scriptContext, RecyclableObject** object)
     {
         Assert(object);
+        if (IsSimdType(aValue))
+        {
+            *object = scriptContext->GetLibrary()->CreateSIMDObject(aValue, JavascriptOperators::GetTypeId(aValue));
+            return TRUE;
+        }
```
Comment (abchatra) [inline lib/Runtime/Language/JavascriptConversion.cpp]: Can you please move this under default? I am not why this should be the first case as other primitive types are more common

Comment (arunetm) [inline lib/Runtime/Language/JavascriptConversion.cpp]: Done

File: lib/Runtime/Language/JavascriptConversion.cpp (hunk 2 above)
Comment (arunetm) [inline lib/Runtime/Language/JavascriptConversion.cpp]: Done

File: lib/Runtime/Language/SimdUtils.h (hunk 1 above)
Comment (arunetm) [inline lib/Runtime/Language/SimdUtils.h]: Will do this in a follow up PR along with the bug fixes. Thanks!

**Comment by arunetm** [comment]
Signing off again after rebasing to the latest master.

**Comment by dilijev** [comment]
Reviewed 17 of 17 files at r3.
Review status: all files reviewed at latest revision, 27 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/325)*

**Pull request merged**